core-bundle/src/DataContainer/DataContainerBuilder.php

```diff
@@ -0,0 +1,195 @@
+<?php
+
+namespace Contao\CoreBundle\DataContainer;
+
+use Contao\CoreBundle\DataContainer\BuilderTemplate\DataContainerBuilderTemplateInterface;
+use Contao\DataContainer;
+use InvalidArgumentException;
+
+class DataContainerBuilder implements DataContainerBuilderInterface
+{
+    private string $name;
+    private array $dca;
+    /** 
+     * @var array<string, DataContainerBuilderTemplateInterface>
+     */
+    private array $templates = [];
+
+    /** 
+     * @param DataContainerBuilderTemplateInterface[] $templates
+     */
+    public function __construct(string $name, iterable $templates = [])
+    {
+        $this->name = $name;
+        $this->dca = $GLOBALS['TL_DCA'][$name] ?? [];
+
+        foreach ($templates as $template) {
+            $this->templates[$template->getName()] = $template;
+        }
+    }
+
+    public function applyTemplate(string ...$templates): self
+    {
+        $merged = [];
+
+        foreach ($templates as $name) {
+            if (!isset($this->templates[$name])) {
+                throw new InvalidArgumentException(sprintf('Could not find template "%s"', $name));
+            }
+
+            $merged = array_replace_recursive($merged, $this->templates[$name]->getConfig());
+        }
+
+        $this->dca = array_replace_recursive($merged, $this->dca);
```

Comment:
```suggestion
        $this->dca = array_replace_recursive($this->dca, $merged);
```
I think this needs to be reversed, otherwise `->applyTemplate('a')->applyTemplate('b')` would merge in a different order than `->applyTemplate('a', 'b')` would.

Also not sure about `array_replace_recursive()` regarding how it handles numeric keys.

Reply:
The thought behind this is that the template shouldn't overwrite any existing DCA defintions. But I guess it would be expected that the template would overwrite the existing definitions, if you apply them _after_ any existing definitions (or later on in the chain).

> Also not sure about array_replace_recursive() regarding how it handles numeric keys.

Yeah, I used it for this quick PoC, but didn't test any side effects.

Reply:
Imho a template shouldn't be applied after. It should only be possible to use templates when you create a DCA yourself (= you define the initial settings). If an extension wants to add e.g. a title to an already existing DCA, it cannot use a template anymore.

Or you'd have to make it pretty advanced by checking if the DCA only consists of templates so far and no manual changes have been made.